src/map_core.rs

```diff
@@ -409,585 +104,291 @@ impl<K, V> Entries for IndexMapCore<K, V> {
-            if let Some((i, hash_proxy)) = self.indices[probe].resolve::<Sz>() {
-                let entry_hash = hash_proxy.get_short_hash(&self.entries, i);
-                // if existing element probed less than us, swap
-                let their_dist = probe_distance(self.mask, entry_hash.into_hash(), probe);
-                if their_dist < dist {
-                    // robin hood: steal the spot if it's better for us
-                    return action.steal(VacantEntry {
-                        map: self,
-                        hash: hash,
-                        key: key,
-                        probe: probe,
-                    });
-                } else if entry_hash == hash && self.entries[i].key == key {
-                    return action.hit(OccupiedEntry {
-                        map: self,
-                        key: key,
-                        probe: probe,
-                        index: i,
-                    });
-                }
-            } else {
-                // empty bucket, insert here
-                return action.empty(VacantEntry {
-                    map: self,
-                    hash: hash,
-                    key: key,
-                    probe: probe,
-                });
-            }
-            dist += 1;
-        });
-    }
-
-    /// phase 2 is post-insert where we forward-shift `Pos` in the indices.
-    fn insert_phase_2<Sz>(&mut self, mut probe: usize, mut old_pos: Pos)
-    where
-        Sz: Size,
+        Q: ?Sized + Equivalent<K>,
     {
-        probe_loop!(probe < self.indices.len(), {
-            let pos = &mut self.indices[probe];
-            if pos.is_none() {
-                *pos = old_pos;
-                break;
-            } else {
-                old_pos = replace(pos, old_pos);
-            }
-        });
+        match self.find_equivalent(hash, key) {
+            Some(raw_bucket) => Some(unsafe { raw_bucket.read() }),
+            None => None,
+        }
     }
 
     pub(crate) fn insert_full(&mut self, hash: HashValue, key: K, value: V) -> (usize, Option<V>)
     where
         K: Eq,
     {
-        self.reserve_one();
-        dispatch_32_vs_64!(self.insert_phase_1::<_>(hash, key, InsertValue(value)))
+        match self.get_index_of(hash, &key) {
+            Some(i) => (i, Some(replace(&mut self.entries[i].value, value))),
+            None => (self.push(hash, key, value), None),
+        }
     }
 
     pub(crate) fn entry(&mut self, hash: HashValue, key: K) -> Entry<K, V>
     where
         K: Eq,
     {
-        self.reserve_one();
-        dispatch_32_vs_64!(self.insert_phase_1::<_>(hash, key, MakeEntry))
-    }
-
-    /// Return probe (indices) and position (entries)
-    pub(crate) fn find_using<F>(&self, hash: HashValue, key_eq: F) -> Option<(usize, usize)>
+        match self.find_equivalent(hash, &key) {
+            Some(raw_bucket) => Entry::Occupied(OccupiedEntry {
+                map: self,
+                raw_bucket,
```

Comment:
I don't think this is very easy to analyze but here we have `map: &'a mut IndexMapCore<K, V>` and a `RawBucket` and the raw bucket is used as a mutable pointer into something that is part of the map, on the surface level this is a borrowing error, and I don't know the rule that would allow this, off hand.

Reply:
On a closer look, `fn shift_remove_bucket(&mut self, raw_bucket: RawBucket)` also runs into the same question, of course.

Reply:
Superficially, the API forces us to do this -- [`erase_no_drop`](https://docs.rs/hashbrown/0.8.0/hashbrown/raw/struct.RawTable.html#method.erase_no_drop) takes `&mut self` and `&Bucket<T>`.

I believe this is sound because the buckets are not directly part of the `RawTable`, but rather part of a separate allocation that the table also holds by raw pointer. It's fine for that to alias the raw pointer in the `Bucket`, and we don't form any references through those pointers. Miri is happy with it, both in `hashbrown`'s CI and when I ran `indexmap`'s testsuite manually under miri.

Reply:
This is just that there is a lot happening in unsafe coding guidelines. Stacked borrows model, rules around raw pointers and pointer provenance. In some cases, a lot more things are forbidden than I thought were with raw pointers. Miri doesn't check raw pointers very well, I think.

I'm sorry, I just don't have a model that I can follow here. That the API forces it is unfortunately not an argument, the other arguments are appreciated 😄  With this code, I feel unmoored and don't know how to think about it - every argument about validity needs to start with the implementation details of RawTable (I can read those), and what do the `&mut self` and `&self` signatures on the methods then even mean anymore?

I'll go incoherent I fear. On a slice it's important that we have `as_mut_ptr(&mut self)` because we cast `&mut [T] as *mut [T] as *mut T`. On a `Vec<T>` is it important we modify through `as_mut_ptr`? Why is there a story that modification through [`Vec::as_ptr -> *const T` is UB](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_ptr)? `RawTable::data_start(&self)` is the same kind of accessor to me.

Reply:
I'm not sure exactly what the rules are about mutating through a `*const T` -- `NonNull` is a notable exception, but maybe this is just an example where `std` can break the rules since it's coupled with the compiler. That is, the result of such UB may be "works as expected" (for now), but if the compiler ever changes (and `NonNull` with it) then that UB could turn nasty to outsiders trying this.

@RalfJung added that warning on slice `as_ptr` (https://github.com/rust-lang/rust/pull/60443) and later copied to `Vec::as_ptr` (https://github.com/rust-lang/rust/pull/61114) -- maybe he can clarify the actual rules?

The [reference](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html) says, "all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an `UnsafeCell<U>`." In that case, I think hashbrown may have a real problem -- e.g. `HashMap::get_mut(&mut self, ...)` calls `RawTable::find(&self, ...)` to get a bucket, and then forms a `&mut V` from that. Does that origin through `&self` render the bucket pointer immutable?

Reply:
`Vec::as_ptr` and `Vec::as_mut_ptr` have the exact same implementation. They both just return the raw pointer held by `RawVec` (notably, via `RawVec::ptr(&self) -> *mut T`), and thus return pointers with the same provenance. The only reason you aren't allowed to mutate through `Vec::as_ptr` is that the docs say so. And I agree that they should; `Vec` is a safe abstraction, so should obey the typical safe rules. `RawVec` is an unsafe abstraction, so obeys the raw pointer rules.

The reference's "data reached through a shared reference" only refers to data behind shared references, and not data behind raw pointers. The important part here is that `RawVec`/`RawTable` (they play the same role here) are raw pointer abstractions.

To be clear, I don't think this understanding of the rules is _explicitly_ condoned by Rust the language. However, a _lot_ of code, both inside and outside std, is written with this interpretation of the rules.

Reply:
> `Vec::as_ptr` and `Vec::as_mut_ptr` have the exact same implementation. They both just return the raw pointer held by `RawVec` (notably, via `RawVec::ptr(&self) -> *mut T`), and thus return pointers with the same provenance.

Peeking at the implementation isn't really informative if that's another "std is special" situation. It tells you that there are no dragons lurking _yet_, but they could hatch in the future. That's why the doc statement is important.

Reply:
> added that warning on slice as_ptr (rust-lang/rust#60443) and later copied to Vec::as_ptr (rust-lang/rust#61114) -- maybe he can clarify the actual rules?

There is no official aliasing model yet, and as such no official rules. But I can explain what [Stacked Borrows](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md) says.

There, the rules are that *which initial "raw borrow" you use* matters, but later the type of the raw ptr does not matter any more. So if `x` is a reference, if you do `x as *mut _` (which is sugar for `&raw mut *x`) then the resulting pointer may be used for mutation. If you do `x as *const _` (sugar for `&raw const *x`), the resulting pointer may *not* be used for mutation.

Also see https://github.com/rust-lang/rust/issues/56604 and in particular [these notes](https://hackmd.io/wpBesJ8PSRG0dX4LHcUZUQ).

Reply:
> The only reason you aren't allowed to mutate through Vec::as_ptr is that the docs say so.

In particular, as usual with library UB, the `Vec` implementation is allowed to change in the future in a way that actually causes language UB when you mutate through `as_ptr`.

Reply:
@RalfJung thanks for commenting. You address the initial raw borrow, but in this case the life of the raw pointer, just comes from the allocator I think? (Just like in Vec, technically)?  In this example, there is never a case where we consider the data pointers to be derived from the `&RawTable` or `&mut RawTable` references at all.

This is the critical piece I think I'm missing - when do we talk about about raw pointers being based on a &/&mut borrow at all?

It seems like in this case, we get a raw pointer from the allocator and we copy the pointer along (we can use an accessor on &RawTable to copy the raw pointer), and it is considered valid to write through this raw pointer all the way through?

Reply:
> This is the critical piece I think I'm missing - when do we talk about about raw pointers being based on a &/&mut borrow at all?

When they are cast from a *reference* (`&T`/`&mut T`).

The initial pointer obtained from the allocator is mutable. It doesnt matter how the raw pointer gets copied around, as long as it does not get turned into a reference and then back into a raw pointer (at that point, the cast rules mentioned above apply).

Reply:
Thanks, previously I've been wondering a lot about what cuviper quotes, the "all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an UnsafeCell<U>."

But it is clarified here that in the accessor for example `RawTable::data_end(&self) -> *mut T` the raw pointer is not considered to be based on the `&RawTable` reference.

In this quote's formulation, it has been unclear what "based on" means and I think that's just due to that it hasn't really been formally specified at all, I know it's an ongoing process. <3

Reply:
Yeah... "based on" is basically the same as [in LLVM](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules).

So, when you compute `p2` from `p1` by doing offsets, then `p2` is "based on" `p1`. But when you *load* `p2` from memory by dereferencing `p1`, that does not induce any kind of "based on" relationship.

Reply:
I don't think the notion of me being confused is fair, when this is a whole bundle of unspecified and not yet formalized rules. From my perspective we're hanging on tight and trying to adapt unsafe code to what the future rules will be, based on ongoing discussion.

Reply:
Sorry, I didn't mean to imply anything negative here.

Pointer "based on" being somehow affected by loads from pointers-to-pointers is a common misconception (but one I did not at all foresee).

Reply:
For pedagogical consideration - for the rest of the community, I mean - I'll mention that I think there is a strong will to have `&mut self/&self` be meaningful, and we are sometimes thinking of this on a Rust conceptual level, not just loading a pointer through a pointer, but want to assign a meaning to accessing through `&mut self` vs `&self`.

Reply:
Fair. OTOH, I think there are also many people that expect this propagation of meaning to stop at raw pointers, and I think this is indeed required to make raw pointers even useful.